nixos/modules/services/web-apps/nextcloud.nix

```diff
@@ -304,6 +304,113 @@ in {
           phone-numbers.
         '';
       };
+
+      objectstore = let
+        s3Arguments = {
+          bucket = mkOption {
+            type = types.str;
+            example = "nextcloud";
+            description = ''
+              The name of the S3 bucket.
+            '';
+          };
+          autocreate = mkOption {
+            type = types.bool;
+            description = ''
+              Create the objectstore if it does not exist.
+            '';
+          };
+          key = mkOption {
+            type = types.str;
+            example = "EJ39ITYZEUH5BGWDRUFY";
+            description = ''
+              The access key for the S3 bucket.
+            '';
+          };
+          secretFile = mkOption {
+            type = types.str;
+            example = "/var/nextcloud-objectstore-s3-secret";
+            description = ''
+              The full path to a file that contains the access secret. Must be
+              readable by user <literal>nextcloud</literal>.
+            '';
+          };
+          hostname = mkOption {
+            type = types.nullOr types.str;
+            default = null;
+            example = "example.com";
+            description = ''
+              Required for some non-Amazon implementations.
+            '';
+          };
+          port = mkOption {
+            type = types.nullOr types.port;
+            default = null;
+            description = ''
+              Required for some non-Amazon implementations.
+            '';
+          };
+          useSsl = mkOption {
+            type = types.nullOr types.bool;
+            default = null;
+            description = ''
+              Use SSL for objectstore access.
+            '';
+          };
+          region = mkOption {
+            type = types.nullOr types.str;
+            default = null;
+            example = "REGION";
+            description = ''
+              Required for some non-Amazon implementations.
+            '';
+          };
+          usePathStyle = mkOption {
+            type = types.bool;
+            default = false;
+            description = ''
+              Required for some non-Amazon S3 implementations.
+
+              Ordinarily, requests will be made with
+              http://bucket.hostname.domain/, but with path style
+              enabled requests are made with
+              http://hostname.domain/bucket instead.
+            '';
+          };
+        };
+      in mkOption {
+        type = types.nullOr (types.submodule {
```

Comment:
I don't see why this is needed. You can just do

``` nix
{
  objectstore = {
    enable = mkEnableOption "...";
    /* all the other objectstore parameters */
  };
}
```

if you can only specify a single object store.

Reply:
The reason why most config is behind `objectstore.s3.*` and not directly behind `objectstore.*` is because nextcloud supports multiple classes of object storage, each with its own set of unique configuration parameters. According to [their docs](https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html) nextcloud currently supports using S3, Openstack Swift V1 and V2 classes of object storage as primary storage. The current structure allows to avoid breaking downstream configurations when someone inevitably decides to add support for Openstack Swift or some other future object storage class that nextcloud might support.

W.r.t. using `nullOr`, I was unsure if you could have a submodule default to `{}` in the case that it has non-default options! I'm happy to remove the `nullOr` indirection and use `{}` as a default instead, it would certainly clean things up a little, though I must admit it's a bit strange to me that nix allows having a submodule type instance that is missing fields that have no defaults. I guess a submodule is just an attrset under the hood though.

Reply:
Then do `objectstore.s3.enable = mkEnableOption "...";`, please. A submodule is absolutely not needed here so far and just causes unnecessary complexity. If this really becomes necessary, we can switch to `types.attrsOf (types.submodule {})` sooner or later or implement the other obejctstore things manually if someone's willing to do that.